File: scripts/acceptance_test.py
from __future__ import annotations

import json
import time
from datetime import datetime, timezone
from pathlib import Path
from typing import Any
from urllib.error import HTTPError
from urllib.request import Request, urlopen

from idea_case_loader import load_idea_case


API = "http://127.0.0.1:8080"
OUTPUT_ROOT = Path(__file__).resolve().parents[1] / "artifacts" / "acceptance"


def request(method: str, path: str, body: dict[str, Any] | None = None, timeout: int = 300) -> Any:
    data = json.dumps(body).encode("utf-8") if body is not None else None
    req = Request(f"{API}{path}", data=data, method=method, headers={"Content-Type": "application/json"})
    try:
        with urlopen(req, timeout=timeout) as response:
            payload = response.read()
            return json.loads(payload) if payload else None
    except HTTPError as exc:
        payload = exc.read().decode("utf-8", errors="replace")
        raise AssertionError(f"{method} {path} returned {exc.code}: {payload}") from exc


def run_direct_acceptance() -> None:
    case = load_idea_case("active-learning-3d")
    results: dict[str, Any] = {"started_at": datetime.now(timezone.utc).isoformat(), "checks": {}}
    health = request("GET", "/api/health")
    assert health["llm"]["provider"] == "openai"
    assert health["llm"]["bridge_required"] is False
    assert health["llm"]["provider_configured"] is True
    assert set(health["llm"]["routing"]["models"]) == {"simple", "medium", "complex"}
    settings = request("GET", "/api/settings/models")
    assert set(settings["tiers"]) == {"simple", "medium", "complex"}
    assert all("key" not in tier for tier in settings["tiers"].values())
    results["checks"]["container_direct_model_settings"] = True

    response = request("POST", "/api/chat", {"message": case.initial_message, "clarification_mode": case.clarification_mode})
    facts = "\n".join(f"{field}: {value}" for field, value in case.confirmed_facts.items())
    facts += "\nThese are explicit user-confirmed facts. Update the whole draft for ProjectSpec review."
    for _ in range(4):
        if response["phase"] == "ready_for_confirmation":
            break
        response = request("POST", "/api/chat", {
            "session_id": response["session_id"],
            "message": facts,
            "clarification_mode": case.clarification_mode,
        })
    assert response["phase"] == "ready_for_confirmation"
    project_id = request("POST", "/api/projects", {"session_id": response["session_id"], "confirmed": True})["project"]["id"]

    deadline = time.time() + 240
    while time.time() < deadline:
        detail = request("GET", f"/api/projects/{project_id}")
        tasks = [task for task in detail["tasks"] if task["kind"] == "research_bootstrap"]
        if tasks and tasks[0]["status"] in {"succeeded", "failed", "cancelled"}:
            break
        time.sleep(2)
    assert tasks and tasks[0]["status"] in {"succeeded", "failed", "cancelled"}

    evidence = request("POST", f"/api/projects/{project_id}/evidence/ingest", {"limit": 3})
    novelty = request("GET", f"/api/projects/{project_id}/novelty")
    assert "related_work" in novelty and "claim_gate" in novelty
    results["project_id"] = project_id
    results["checks"].update({
        "evidence_first_related_work": {"stored_count": evidence["stored_count"], "assessment": novelty["assessment"]},
        "topic_plan_execution": {
            "executed": False,
            "approval_required": True,
            "model_request_not_started": True,
            "note": "The acceptance entry does not call the model-backed topic planner or substitute an unrelated experiment.",
        },
    })

    paused = request("POST", f"/api/projects/{project_id}/state", {"action": "pause", "reason": "Acceptance state gate"})
    resumed = request("POST", f"/api/projects/{project_id}/state", {"action": "resume", "reason": "Acceptance state gate complete"})
    assert paused["status"] == "paused" and resumed["status"] == "active"
    results["checks"]["project_state_gate"] = True
    results["finished_at"] = datetime.now(timezone.utc).isoformat()
    OUTPUT_ROOT.mkdir(parents=True, exist_ok=True)
    target = OUTPUT_ROOT / f"acceptance-{datetime.now().strftime('%Y%m%d-%H%M%S')}.json"
    target.write_text(json.dumps(results, ensure_ascii=False, indent=2), encoding="utf-8")
    print(json.dumps({"status": "passed", "result_file": str(target), **results}, ensure_ascii=False, indent=2))


if __name__ == "__main__":
    run_direct_acceptance()

File: scripts/acceptance-test.ts
import { existsSync, mkdirSync, writeFileSync } from 'node:fs'
import { resolve } from 'node:path'
import { loadIdeaCase, repositoryRoot } from './idea-case-loader.js'

const apiBase = process.env.RESEARCH_API_URL || 'http://127.0.0.1:8080'
const envPath = resolve(repositoryRoot, '.env')
if (existsSync(envPath) && typeof process.loadEnvFile === 'function') process.loadEnvFile(envPath)
async function request<T>(method: string, path: string, body?: unknown): Promise<T> {
  const response = await fetch(`${apiBase}${path}`, { method, headers: { 'content-type': 'application/json' }, ...(body === undefined ? {} : { body: JSON.stringify(body) }), signal: AbortSignal.timeout(300_000) })
  const payload = await response.json().catch(() => ({})) as T & { code?: string; message?: string }
  if (!response.ok) throw new Error(`${method} ${path}: ${response.status} ${payload.code || ''} ${payload.message || ''}`)
  return payload
}

const testCase = loadIdeaCase(process.env.RESEARCH_IDEA_CASE || 'active-learning-3d')
const health = await request<Record<string, unknown>>('GET', '/api/health')
if (health.runtime !== 'native-typescript' || health.container_runtime_required !== false) throw new Error('native runtime health contract failed')
const settings = await request<{ tiers: Record<string, Record<string, unknown>> }>('GET', '/api/settings/models')
if (Object.values(settings.tiers).some(tier => 'key' in tier)) throw new Error('model settings exposed a key')
if (Object.values(settings.tiers).some(tier => tier.key_configured !== true)) throw new Error('one or more model tiers have no configured key')
for (const [tier, suffix] of [['simple', 'SIMPLE'], ['medium', 'MEDIUM'], ['complex', 'COMPLEX']] as const) {
  const expectedUrl = process.env[`RESEARCH_MODEL_URL_${suffix}`]
  if (!expectedUrl || settings.tiers[tier]?.url !== expectedUrl) throw new Error(`${tier} model URL does not match project .env`)
}
let chat = await request<Record<string, any>>('POST', '/api/chat', { message: testCase.initial_message, clarification_mode: testCase.clarification_mode })
const routes = [{ tier: chat.model_tier, model: chat.model, reasoning_effort: chat.reasoning_effort }]
const facts = `${Object.entries(testCase.confirmed_facts).map(([key, value]) => `${key}: ${value}`).join('\n')}\nThese are explicit user-confirmed facts.`
for (let index = 0; index < 4 && chat.phase !== 'ready_for_confirmation'; index += 1) {
  chat = await request('POST', '/api/chat', { session_id: chat.session_id, message: facts, clarification_mode: testCase.clarification_mode })
  routes.push({ tier: chat.model_tier, model: chat.model, reasoning_effort: chat.reasoning_effort })
}
if (chat.phase !== 'ready_for_confirmation') throw new Error('Idea clarification did not reach confirmation')
const suffix = crypto.randomUUID().replaceAll('-', '').slice(0, 4)
const project = await request<{ project_id: string }>('POST', '/api/projects', {
  slug: `native-acceptance-${suffix}`,
  title: testCase.description,
})
const paused = await request<{ status: string }>('POST', `/api/projects/${project.project_id}/state`, { action: 'pause', reason: 'Native acceptance state gate' })
const resumed = await request<{ status: string }>('POST', `/api/projects/${project.project_id}/state`, { action: 'resume', reason: 'Native acceptance state gate complete' })
if (paused.status !== 'paused' || resumed.status !== 'active') throw new Error('project state gate failed')
const result = { status: 'passed', runtime: 'native-typescript', project_id: project.project_id, idea_case: testCase.id, completed_at: new Date().toISOString(), checks: { model_keys_hidden: true, model_urls_match_env: true, real_model_calls: routes.length, project_state_gate: true }, routes }
const output = resolve(repositoryRoot, 'artifacts', 'acceptance')
mkdirSync(output, { recursive: true })
const target = resolve(output, `acceptance-${new Date().toISOString().replace(/[-:TZ.]/g, '').slice(0, 14)}.json`)
writeFileSync(target, `${JSON.stringify(result, null, 2)}\n`)
console.log(JSON.stringify({ ...result, result_file: target }, null, 2))
